### ISSUE TITLE
Memory Issue/Leak (RAM piles up infinitely)

### ISSUE THREAD

**Issue by brr53** [ultralytics/ultralytics#6259]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

When training a model with Yolov8, after the training completes, the RAM is not released. 
This becomes an issue when training multiple models iteratively in Python code due to RAM stacking until 99.99%

In my testing I wait for each model to finish training before moving onto the next. I do not use any parallelism. Yet RAM goes to 99.99% by the 4th model training. 

How can we free memory after training completes? 

The current hack I use is to run each model training as its own process and to kill it once training is done. This hack was also suggested by a Ultralytics Community Moderator. But surely this can't be the correct way to handle things?

### Environment

Ultralytics YOLOv8.0.196 🚀 Python-3.9.16 torch-2.0.1+cu117 CUDA:0 (Tesla T4, 15984MiB)

RAM 64GB
Python 3.10
Tested centos, ubuntu, windows

### Minimal Reproducible Example

Train a Yolov8 models.

model.train(task=task, data=data_yaml_path, epochs=5, imgsz=imgsz,
            single_cls=True, cache=False, batch=16, workers=1,
            device=device_list) 

### Additional

Community helper said he suspects Yolo keeps everything in memory until session is closed. But I don't think it is satisfactory to resolve RAM leak issue by restarting the program.

Image after 5 epoch train completes (COCO 128) 
Memory is not freed after completion.
![image](https://github.com/ultralytics/ultralytics/assets/60824624/8a5b070c-dfe4-4235-9a03-376997a52dc0)

Image of that 5 epoch train program on exit
Memory is released
![image](https://github.com/ultralytics/ultralytics/assets/60824624/9e7ff711-adb6-40e3-a271-a3a3c06ff7f0)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@brr53 your minimum reproducible example trains only 1 model, so it's not complete for us to reproduce your issue. Can you please supply a complete MRE we can use?

**Comment by brr53** [comment]
@glenn-jocher even when just training 1 model. You can see after the training is complete the RAM/memory used during training is not released. You can just train a model for 5 epochs using COCO128 for example.

**Comment by glenn-jocher** [comment]
@brr53 thank you for the clarification. The behavior you're encountering is indeed not the intended experience. Memory should be released after the training process is complete, and residual consumption could indicate that some resources, perhaps tensors or computational graphs, might not be properly freed.

While we look into this, you might consider a programmatic workaround by encapsulating the training routine within a separate function or subprocess, which should have its memory completely released when the function exits or the subprocess is terminated. However, I understand this is similar to the workaround you are currently employing.

We appreciate that you're willing to submit a PR. Before you proceed, I would encourage you to open a new issue with the detailed description of the problem, reproduction steps, and your environment information. It would be beneficial if you can profile the memory allocation and identify the potential source of the memory leak during training to make your contribution more substantial.

For the time being, you can manage the issue with the mentioned subprocess method or monitor and manually clear the caches with torch.cuda memory management tools, in case they are contributing to the issue. If you find specific areas in the codebase that are leading to this behavior, please reference them in the issue report.

We'll prioritize examining the issue you've raised as it's critical for those needing to train multiple models sequentially. Further discussion and collaboration with the community developers might be necessary to resolve it.

**Comment by Luchen-077** [comment]
> @brr53 thank you for the clarification. The behavior you're encountering is indeed not the intended experience. Memory should be released after the training process is complete, and residual consumption could indicate that some resources, perhaps tensors or computational graphs, might not be properly freed.
> 
> While we look into this, you might consider a programmatic workaround by encapsulating the training routine within a separate function or subprocess, which should have its memory completely released when the function exits or the subprocess is terminated. However, I understand this is similar to the workaround you are currently employing.
> 
> We appreciate that you're willing to submit a PR. Before you proceed, I would encourage you to open a new issue with the detailed description of the problem, reproduction steps, and your environment information. It would be beneficial if you can profile the memory allocation and identify the potential source of the memory leak during training to make your contribution more substantial.
> 
> For the time being, you can manage the issue with the mentioned subprocess method or monitor and manually clear the caches with torch.cuda memory management tools, in case they are contributing to the issue. If you find specific areas in the codebase that are leading to this behavior, please reference them in the issue report.
> 
> We'll prioritize examining the issue you've raised as it's critical for those needing to train multiple models sequentially. Further discussion and collaboration with the community developers might be necessary to resolve it.

Hello, I wonder if the problem has been solved recently, I meet the same problem when I train YOLOv11n, RAM piles up infinitely like this. Cleaning the cache every epoch will significantly slow down the training speed. I don't know how to solve it exactly.

**Comment by Y-T-G** [comment]
Should be fixed in latest Ultralytics
